File: agent_service/models/agent_queue.py
"""Persistent user-owned Agent task queue records."""

from __future__ import annotations

from datetime import datetime

from sqlmodel import Field, SQLModel

from agent_service.core.agent_config import DEFAULT_BUSINESS_LIMITS

from agent_service.models.session import utc_now


class AgentQueueTaskRecord(SQLModel, table=True):
    """One independently runnable Agent task and its dedicated session."""

    __tablename__ = "agent_queue_tasks"

    task_id: str = Field(primary_key=True, max_length=DEFAULT_BUSINESS_LIMITS.standard_id_max_length)
    user_id: str = Field(index=True, max_length=DEFAULT_BUSINESS_LIMITS.medium_name_max_length)
    prompt: str
    priority: str = Field(default="medium", max_length=DEFAULT_BUSINESS_LIMITS.short_status_max_length, index=True)
    status: str = Field(default="pending", max_length=DEFAULT_BUSINESS_LIMITS.short_status_max_length, index=True)
    session_id: str | None = Field(default=None, index=True, max_length=DEFAULT_BUSINESS_LIMITS.standard_id_max_length)
    attachments_json: str = Field(default="[]")
    previous_task_id: str | None = Field(default=None, index=True, max_length=DEFAULT_BUSINESS_LIMITS.standard_id_max_length)
    started_at: datetime | None = None
    finished_at: datetime | None = None
    terminated_at: datetime | None = None
    created_at: datetime = Field(default_factory=utc_now, index=True)
    updated_at: datetime = Field(default_factory=utc_now, index=True)


class AgentQueueSettingsRecord(SQLModel, table=True):
    """Per-user queue concurrency preference."""

    __tablename__ = "agent_queue_settings"

    user_id: str = Field(primary_key=True, max_length=DEFAULT_BUSINESS_LIMITS.medium_name_max_length)
    max_concurrency: int = Field(
        default=DEFAULT_BUSINESS_LIMITS.agent_queue_default_concurrency,
        ge=DEFAULT_BUSINESS_LIMITS.nonempty_min_length,
        le=DEFAULT_BUSINESS_LIMITS.agent_queue_max_concurrency,
    )
    updated_at: datetime = Field(default_factory=utc_now)

File: editor/src/api/agentQueue.ts
/**
 * Durable Agent task queue API.
 *
 * Usage:
 * The queue store and board use these functions exclusively; no queue state is
 * fabricated in the browser.
 */

import { apiDelete, apiGet, apiPost, apiPut } from '@/api/client'
import { API_ROUTES } from '@/router/api_routes'
import type { AgentUploadedAttachment } from '@/stores/chat'

export type AgentQueuePriority = 'critical' | 'high' | 'medium' | 'low' | 'whenever'
export type AgentQueueStatus = 'pending' | 'running' | 'review' | 'confirmed' | 'terminated'

export interface AgentQueueTask {
  task_id: string
  user_id: string
  prompt: string
  priority: AgentQueuePriority
  status: AgentQueueStatus
  session_id: string | null
  attachments: AgentUploadedAttachment[]
  started_at: string | null
  finished_at: string | null
  terminated_at: string | null
  created_at: string
  updated_at: string
  previous_task_id: string | null
  task_list?: { items?: Array<{ status: string }> } | null
}

export interface QueueTaskPayload {
  user_id: string
  prompt: string
  priority: AgentQueuePriority
  attachments: AgentUploadedAttachment[]
}

/** Load the live board or the confirmed/terminated history for one user. */
export function fetchQueue(userId: string, history = false) {
  return apiGet<{ tasks: AgentQueueTask[]; settings: { max_concurrency: number } }>(
    API_ROUTES.AGENT_QUEUE_TASKS,
    { user_id: userId, history },
  )
}

/** Persist a new independently executable Agent task. */
export function createQueueTask(payload: QueueTaskPayload & { session_id: string }) {
  return apiPost<AgentQueueTask>(API_ROUTES.AGENT_QUEUE_TASKS, payload)
}

/** Replace the editable fields of a pending task. */
export function updateQueueTask(taskId: string, payload: QueueTaskPayload) {
  return apiPut<AgentQueueTask>(API_ROUTES.AGENT_QUEUE_TASK(taskId), payload)
}

/** Transition a task to confirmed or terminated. */
export function transitionQueueTask(taskId: string, userId: string, status: 'confirmed' | 'terminated') {
  return apiPost<AgentQueueTask>(API_ROUTES.AGENT_QUEUE_TRANSITION(taskId), { user_id: userId, status })
}

/** Persist the per-user concurrent worker limit. */
export function updateQueueSettings(userId: string, maxConcurrency: number) {
  return apiPut<{ max_concurrency: number }>(API_ROUTES.AGENT_QUEUE_SETTINGS, {
    user_id: userId,
    max_concurrency: maxConcurrency,
  })
}

/** Delete an unclaimed task. */
export function deleteQueueTask(taskId: string, userId: string) {
  return apiDelete<{ deleted: boolean }>(API_ROUTES.AGENT_QUEUE_TASK(taskId), { user_id: userId })
}

/** Return a reviewed task to pending with a replacement prompt and attachments. */
export function continueQueueTask(taskId: string, payload: Pick<QueueTaskPayload, 'user_id' | 'prompt' | 'attachments'>) {
  return apiPost<AgentQueueTask>(API_ROUTES.AGENT_QUEUE_CONTINUE(taskId), payload)
}
